platforms/cuda/src/CudaTorchKernels.cpp

```diff
@@ -49,7 +49,13 @@ if (result != CUDA_SUCCESS) { \
     throw OpenMMException(m.str());\
 }
 
+CudaCalcTorchForceKernel::CudaCalcTorchForceKernel(string name, const Platform& platform, CudaContext& cu) :
+        CalcTorchForceKernel(name, platform), hasInitializedKernel(false), cu(cu) {
+    CHECK_RESULT(cuDevicePrimaryCtxRetain(&primaryContext, cu.getDevice()), "Failed to retain the primary context");
```

Comment:
I think this is incorrect in a couple of ways.  First, there's no guarantee that PyTorch will have set up the primary context yet at this point.  Inside `initialize()` we ensure it has been initialized with the line

https://github.com/openmm/openmm-torch/blob/661b004eb1a2b1c9cca8c8c857d3336f17762017/platforms/cuda/src/CudaTorchKernels.cpp#L64

Depending on what else has happened earlier, it might have already been initialized at this point but it's not guaranteed.

Second, by recording the primary context here you're assuming it will be created once and never changed.  That's the typical pattern with the runtime API, but it isn't guaranteed.  `cudaDeviceReset()` can be used to delete the primary context and create a new one.  We shouldn't make assumptions about what PyTorch might do.  The safe approach is to query the primary context immediately before pushing it with `cuCtxPushCurrent()`.

Reply:
Regarding the second point, `cudaDeviceReset` (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1gef69dd5c6d0206c2b8d099abac61f217) destroys all the contexts, not only the primary one. Also, all the memory allocations are wiped out. So, it is game over, the only thing OpenMM-Torch can do is to crash.

Reply:
Regarding the first point, the primary context will be activate in the constructor by `cuDevicePrimaryCtxRetain` (if not yet active) (https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__PRIMARY__CTX.html#group__CUDA__PRIMARY__CTX_1g9051f2d5c31501997a6cb0530290a300) and it is guaranteed to be active until `cuDevicePrimaryCtxRelease` is called in the destructor.

Reply:
> cudaDeviceReset destroys all the contexts, not only the primary one.

I don't think that's correct.  The documentation you linked refers to the runtime API, which treats "device" and "context" as synonymous.  Calling methods on the runtime API doesn't affect resources allocated separately by the lower level driver API.

Reply:
This documentation in the [CUDA Programming Guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#initialization) is a bit more explicit about it:

> When a host thread calls cudaDeviceReset(), this destroys the primary context of the device the host thread currently operates on (i.e., the current device as defined in [Device Selection](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-selection)). The next runtime function call made by any host thread that has this device as current will create a new primary context for this device.

Reply:
Even if it destroy just the primary context, it is still gave over. PyTorch stores the NNP model on the primary context and it will be gone.